File: ruby/spanish_converter.rb
require 'sinatra'
require 'net/http'
require 'json'
require 'uri'

set :bind, '127.0.0.1'
set :port, 4567

WSDL_URL = 'http://www.dataaccess.com/webservicesserver/NumberConversion.wso'

NUMBER_WORDS = {
  0 => 'zero', 1 => 'one', 2 => 'two', 3 => 'three', 4 => 'four',
  5 => 'five', 6 => 'six', 7 => 'seven', 8 => 'eight', 9 => 'nine',
  10 => 'ten', 11 => 'eleven', 12 => 'twelve', 13 => 'thirteen', 14 => 'fourteen',
  15 => 'fifteen', 16 => 'sixteen', 17 => 'seventeen', 18 => 'eighteen', 19 => 'nineteen',
  20 => 'twenty', 30 => 'thirty', 40 => 'forty', 50 => 'fifty',
  60 => 'sixty', 70 => 'seventy', 80 => 'eighty', 90 => 'ninety'
}.freeze

def local_number_to_words(n)
  return NUMBER_WORDS[n] if NUMBER_WORDS.key?(n)
  return "twenty-#{NUMBER_WORDS[n - 20]}" if n < 30

  if n < 100
    tens = (n / 10) * 10
    units = n % 10
    return units.zero? ? NUMBER_WORDS[tens] : "#{NUMBER_WORDS[tens]} #{NUMBER_WORDS[units]}"
  end

  'number out of range'
end

def soap_to_english(n)
  soap_body = <<~XML
    <?xml version="1.0" encoding="utf-8"?>
    <soap:Envelope xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/">
      <soap:Body>
        <NumberToWords xmlns="http://www.dataaccess.com/webservicesserver/">
          <ubiNum>#{n}</ubiNum>
        </NumberToWords>
      </soap:Body>
    </soap:Envelope>
  XML

  uri = URI.parse(WSDL_URL)
  http = Net::HTTP.new(uri.host, uri.port)
  request = Net::HTTP::Post.new(uri.request_uri)
  request['Content-Type'] = 'text/xml; charset=utf-8'
  request['SOAPAction'] = 'http://www.dataaccess.com/webservicesserver/NumberToWords'
  request.body = soap_body

  response = http.request(request)
  return local_number_to_words(n) unless response.is_a?(Net::HTTPSuccess)

  if response.body =~ %r{<NumberToWordsResult>(.*?)</NumberToWordsResult>}m
    return Regexp.last_match(1).strip
  end

  if response.body =~ %r{<m:NumberToWordsResult>(.*?)</m:NumberToWordsResult>}m
    return Regexp.last_match(1).strip
  end

  local_number_to_words(n)
end

def translate_text(english)
  uri = URI.parse("https://translate.googleapis.com/translate_a/single?client=gtx&sl=en&tl=es&dt=t&q=#{URI.encode_www_form_component(english)}")
  response = Net::HTTP.get_response(uri)
  return english if !response.is_a?(Net::HTTPSuccess)

  data = JSON.parse(response.body)
  data.dig(0, 0, 0)
end

get '/' do
  n = (params['n'] || '10').to_i
  english = begin
    soap_to_english(n)
  rescue Net::OpenTimeout, Net::ReadTimeout, SocketError, StandardError
    local_number_to_words(n)
  end

  translate_text(english) || english
end

File: rust/src/bin/spanish_converter.rs
use reqwest::blocking::Client;
use serde_json::Value;
use tiny_http::{Response, Server};
use std::error::Error;

fn soap_to_english(n: i32) -> Result<String, Box<dyn Error>> {
    let wsdl_url = "https://www.dataaccess.com/webservicesserver/NumberConversion.wso";
    let body = format!(r#"<?xml version="1.0" encoding="utf-8"?>
<soap:Envelope xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/">
  <soap:Body>
    <NumberToWords xmlns="http://www.dataaccess.com/webservicesserver/">
      <ubiNum>{}</ubiNum>
    </NumberToWords>
  </soap:Body>
</soap:Envelope>"#, n);

    let client = Client::new();
    let resp = client.post(wsdl_url)
        .header("Content-Type", "text/xml; charset=utf-8")
        .header("SOAPAction", "http://www.dataaccess.com/webservicesserver/NumberToWords")
        .body(body)
        .send()?;

    let text = resp.text()?;
    if let Some(start) = text.find("<NumberToWordsResult>") {
        let start = start + "<NumberToWordsResult>".len();
        let end = text.find("</NumberToWordsResult>").ok_or("missing end tag")?;
        return Ok(text[start..end].trim().to_string());
    }

    if let Some(start) = text.find("<m:NumberToWordsResult>") {
        let start = start + "<m:NumberToWordsResult>".len();
        let end = text.find("</m:NumberToWordsResult>").ok_or("missing end tag")?;
        return Ok(text[start..end].trim().to_string());
    }

    Ok(local_number_to_words(n))
}

fn translate(text: &str) -> Result<String, Box<dyn Error>> {
    let url = format!(
        "https://translate.googleapis.com/translate_a/single?client=gtx&sl=en&tl=es&dt=t&q={}",
        urlencoding::encode(text)
    );
    let resp = Client::new().get(&url).send()?.text()?;
    let parsed: Value = serde_json::from_str(&resp)?;
    let translated = parsed[0][0][0].as_str().unwrap_or("");
    if translated.is_empty() {
        Ok(english_to_spanish(text))
    } else {
        Ok(translated.to_string())
    }
}

fn local_number_to_words(n: i32) -> String {
    let words = [
        "zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine", "ten",
        "eleven", "twelve", "thirteen", "fourteen", "fifteen", "sixteen", "seventeen", "eighteen", "nineteen", "twenty",
    ];

    if (0..=20).contains(&n) {
        return words[n as usize].to_string();
    }

    if n < 30 {
        return format!("twenty {}", words[(n - 20) as usize]);
    }

    if n < 100 {
        let tens_words = ["zero", "ten", "twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety"];
        let tens = (n / 10) as usize;
        let units = (n % 10) as usize;
        return if units == 0 {
            tens_words[tens].to_string()
        } else {
            format!("{} {}", tens_words[tens], words[units])
        };
    }

    "number out of range".to_string()
}

fn english_to_spanish(text: &str) -> String {
    match text.trim().to_lowercase().as_str() {
        "zero" => "cero".to_string(),
        "one" => "uno".to_string(),
        "two" => "dos".to_string(),
        "three" => "tres".to_string(),
        "four" => "cuatro".to_string(),
        "five" => "cinco".to_string(),
        "six" => "seis".to_string(),
        "seven" => "siete".to_string(),
        "eight" => "ocho".to_string(),
        "nine" => "nueve".to_string(),
        "ten" => "diez".to_string(),
        "eleven" => "once".to_string(),
        "twelve" => "doce".to_string(),
        "thirteen" => "trece".to_string(),
        "fourteen" => "catorce".to_string(),
        "fifteen" => "quince".to_string(),
        "sixteen" => "dieciséis".to_string(),
        "seventeen" => "diecisiete".to_string(),
        "eighteen" => "dieciocho".to_string(),
        "nineteen" => "diecinueve".to_string(),
        "twenty" => "veinte".to_string(),
        other if other.starts_with("twenty ") => {
            let suffix = &other[7..];
            format!("veinti{}", english_to_spanish(suffix))
        }
        other if other.starts_with("thirty ") => format!("treinta y {}", english_to_spanish(&other[7..])),
        other if other.starts_with("forty ") => format!("cuarenta y {}", english_to_spanish(&other[6..])),
        other if other.starts_with("fifty ") => format!("cincuenta y {}", english_to_spanish(&other[6..])),
        other if other.starts_with("sixty ") => format!("sesenta y {}", english_to_spanish(&other[6..])),
        other if other.starts_with("seventy ") => format!("setenta y {}", english_to_spanish(&other[8..])),
        other if other.starts_with("eighty ") => format!("ochenta y {}", english_to_spanish(&other[7..])),
        other if other.starts_with("ninety ") => format!("noventa y {}", english_to_spanish(&other[7..])),
        _ => text.to_string(),
    }
}

fn extract_query_n(query: &str) -> i32 {
    for part in query.split('&') {
        let mut kv = part.splitn(2, '=');
        if let (Some(key), Some(value)) = (kv.next(), kv.next()) {
            if key == "n" {
                if let Ok(parsed) = value.parse::<i32>() {
                    return parsed;
                }
            }
        }
    }
    10
}

fn main() -> Result<(), Box<dyn Error + Send + Sync>> {
    let server = Server::http("0.0.0.0:8080")?;
    println!("Servidor iniciado en http://localhost:8080");

    for request in server.incoming_requests() {
        let query = request.url().split('?').nth(1).unwrap_or("");
        let n = extract_query_n(query);

        let english = soap_to_english(n).unwrap_or_else(|_| local_number_to_words(n));
        let spanish = translate(&english).unwrap_or_else(|_| english_to_spanish(&english));

        let response = Response::from_string(spanish);
        let _ = request.respond(response);
    }

    Ok(())
}
